### PR TITLE
Add additional tensor operations

### LINKED ISSUE #15406: Add other methods to RankTwo and RankFour Tensors
## Reason
In tensor mechanics, sometimes we need methods that are not available in the `RankTwoTensor` but are commonly used.

## Design
add some methods and unit tests

## Impact
Ideally, this PR will allow consolidating and get rid of duplicated code in tensor mechanics app

### PR DESCRIPTION
**Pull request by arovinelli**
closes #15406 

## Reason
needed by #17157

## Design
add methods to rank two and rank four tenors

## Impact
more tensor operations, also this needs to go in before I can complete #17157

### CODE REVIEW

File: framework/include/utils/RankFourTensorImplementation.h [hunk 1]
```diff
@@ -455,6 +456,51 @@ RankFourTensorTempl<T>::transposeMajor() const
   return result;
 }
 
+template <typename T>
+RankFourTensorTempl<T>
+RankFourTensorTempl<T>::transposeIJ() const
```
Comment (bwspenc) [inline framework/include/utils/RankFourTensorImplementation.h:461]: Should this be `transposeIj` with our naming conventions for these methods?

**Review by bwspenc — COMMENTED** [review]
Aside from my question about naming convention, this looks good to me.

File: framework/include/utils/RankFourTensorImplementation.h (hunk 1 above)
Comment (arovinelli) [reply framework/include/utils/RankFourTensorImplementation.h]: Thanks for the review @bwspenc . You are right, I updated the name as suggested

**Comment by dschwen** [comment]
@matthiasneuner made a comment in our Discussions section: https://github.com/idaholab/moose/discussions/18194 where he mentions Fastor, a Tensor library that supports template based Einstein summation and would generalize all methods added here.

**Comment by bwspenc** [comment]
@dschwen I believe that adopting Fastor would be a pretty major undertaking because we would need to replace our RankXTensor classes with Fastor equivalents. While that might be worth considering, that would be more of a long-term thing that shouldn't really affect this PR, because this adds functionality that he needs now. Would you agree?

**Comment by arovinelli** [comment]
@dschwen I would definitely prefer to merge this now and adapt all the code to Fastor later instead of waiting for Fastor to be merged in. Also letting this in would simplify keeping track of all the operations we need to update with Fastor. Still, I'm a fan of einstein summation

**Comment by dschwen** [comment]
> @dschwen I believe that adopting Fastor would be a pretty major undertaking because we would need to replace our RankXTensor classes with Fastor equivalents. While that might be worth considering, that would be more of a long-term thing that shouldn't really affect this PR, because this adds functionality that he needs now. Would you agree?

No. As I commented TensorMap would allow us to get a Factor view of our existing tensor classes.

I'm fine with this getting merged, but I foresee these methods to become deprecated in the not too distant future.

**Comment by arovinelli** [comment]
@dschwen I'll take responsibility for updating my code to Fastor as soon as it gets in.

**Comment by arovinelli** [comment]
@bwspenc so can we merge this?

**Review by dschwen — DISMISSED** [review]

**Comment by arovinelli** [comment]
wait, I need to add another method -.- einstein summation is annoying. I'm really looking forward for Fastor

**Review by bwspenc — APPROVED** [review]

**Pull request merged**